src/robot/Robot.cpp

```diff
@@ -1,21 +1,454 @@
 #include "aikido/robot/Robot.hpp"
 
+#include <dart/collision/fcl/FCLCollisionDetector.hpp>
+#include <dart/common/Console.hpp>
+
+#include "aikido/constraint/TestableIntersection.hpp"
+#include "aikido/planner/SnapConfigurationToConfigurationPlanner.hpp"
+#include "aikido/planner/dart/ConfigurationToConfiguration.hpp"
+#include "aikido/planner/dart/ConfigurationToConfiguration_to_ConfigurationToConfiguration.hpp"
+#include "aikido/planner/dart/ConfigurationToConfiguration_to_ConfigurationToTSR.hpp"
+#include "aikido/planner/dart/ConfigurationToTSR.hpp"
+#include "aikido/robot/util.hpp"
+#include "aikido/statespace/GeodesicInterpolator.hpp"
+#include "aikido/statespace/StateSpace.hpp"
+#include "aikido/statespace/dart/MetaSkeletonStateSpace.hpp"
+
 namespace aikido {
 namespace robot {
 
 //==============================================================================
-dart::dynamics::MetaSkeletonPtr Robot::getMetaSkeleton()
+Robot::Robot(
+    const dart::dynamics::MetaSkeletonPtr& skeleton,
+    const std::string name,
+    const aikido::control::TrajectoryExecutorPtr& trajExecutor)
+  : mName(name)
+  , mMetaSkeleton(skeleton)
+  , mDofs(util::dofNamesFromSkeleton(skeleton))
+  , mCollisionDetector(dart::collision::FCLCollisionDetector::create())
+{
+  mStateSpace = std::make_shared<statespace::dart::MetaSkeletonStateSpace>(
+      mMetaSkeleton.get());
+  auto skeletonObj = mMetaSkeleton->getBodyNode(0)->getSkeleton();
+  skeletonObj->enableSelfCollisionCheck();
+  skeletonObj->disableAdjacentBodyCheck();
+
+  setTrajectoryExecutor(trajExecutor);
+  auto rngSeed = std::chrono::system_clock::now().time_since_epoch().count();
+  mRng = std::make_unique<common::RNGWrapper<std::default_random_engine>>(
+      rngSeed);
+  mWorld = std::make_shared<aikido::planner::World>();
+  mWorld->addSkeleton(skeletonObj);
+}
+
+//==============================================================================
+void Robot::ignoreSelfCollisionPairs(
+    const std::vector<std::pair<std::string, std::string>> bodyNodeList)
 {
-  return std::const_pointer_cast<dart::dynamics::MetaSkeleton>(
-      const_cast<const Robot*>(this)->getMetaSkeleton());
+  for (auto pair : bodyNodeList)
+  {
+    auto body0 = mMetaSkeleton->getBodyNode(pair.first);
+    auto body1 = mMetaSkeleton->getBodyNode(pair.second);
+    if (body0 && body1)
+    {
+      mSelfCollisionFilter->addBodyNodePairToBlackList(body0, body1);
+    }
+  }
 }
 
 //==============================================================================
-statespace::dart::MetaSkeletonStateSpacePtr Robot::getStateSpace()
+std::future<void> Robot::executeTrajectory(
+    const trajectory::TrajectoryPtr& trajectory) const
+{
+  if (!mTrajectoryExecutor)
+  {
+    throw std::runtime_error(
+        "TrajectoryExecutor is null, cannot execute trajectory.");
+  }
+
+  // Ensure all other executors that could contain these joints are cancelled
+  if (mRootRobot)
+  {
+    mRootRobot->cancelAllTrajectories(false, true);
```

Comment:
Why don't you just run this with true, true? It seems like you are replicating the logic since you're looping over all subrobots anyway